FreeRTOS/Test/Target/tests/smp/template/test_name.c

```diff
@@ -0,0 +1,97 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://github.com/FreeRTOS
+ *
+ */
+
+/**
+ * @file test.c
+ * @brief Describe the test here briefly.
+ *
+ * Describe the test here in detail.
+ */
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+/* Unity testing includes. */
+#include "unity.h"
+
+/*-----------------------------------------------------------*/
+
+#if ( configNUMBER_OF_CORES < 2 )
+    #error This test is for FreeRTOS SMP and therefore, requires at least 2 cores.
+#endif
+
+#ifndef TEST_CONFIG_H
+    #error test_config.h must be included at the end of FreeRTOSConfig.h.
+#endif
+
+/*-----------------------------------------------------------*/
+
+/* Declare #defines used in this file here. */
+
+/*-----------------------------------------------------------*/
+
+/* Declare local functions used in this file here. */
+
+/*-----------------------------------------------------------*/
+
+/* Declare local variables used in this file here. */
+
+/*-----------------------------------------------------------*/
+
+/* Runs before every test, put init calls here. */
+void setUp( void )
+{
+    /* Create FreeRTOS resources required for the test. */
+}
+/*-----------------------------------------------------------*/
+
+/* Runs after every test, put clean-up calls here. */
+void tearDown( void )
+{
+    /* Delete all the FreeRTOS resources created in setUp. */
+}
+/*-----------------------------------------------------------*/
+
+void Test_TestCaseName( void )
+{
+    /* Peform any API call needed for the test. */
+
+    /* Verify the result. */
+}
+/*-----------------------------------------------------------*/
+
+/* Function that runs the test case. This function must be called
+ * from a FreeRTOS task. */
+void vRunTestCaseName( void )
+{
+    UNITY_BEGIN();
+
+    RUN_TEST( Test_TestCaseName );
+
+    UNITY_END();
```

Comment:
I've noticed that each test case must individually call `UNITY_BEGIN()`/`UNITY_END()`. Is the intention here to allow each test case to act as a stand alone unity test application? (i.e., each test case can be individually compiled to a standalone test app).

Would it be possible to delegate the calling of `UNITY_BEGIN()`/`RUN_TEST()`/`UNITY_END()` to the target instead (i.e., called from the target's `prvTestRunnerTask()`)? From our experience, doing this will make integrating unity tests into a CI pipeline easier (where tests run on a physical boards and are triggered by a CI runner).

- Building is usually one of the more time consuming parts of a CI pipeline. Thus allowing all test cases to be put into the same test app (i.e., same binary) would reduce the time cost of each CI pipeline.
- Likewise, having a single test app contain all the tests would mean the runner only needs to flash the board once
- The test app will only need to call `UNITY_BEGIN()`/`UNITY_END()` once at the start/end of the test applications life time.
- If the test app runs multiple tests (via a series/loop of `RUN_TEST()` calls)...
  - The test app can continue running other tests even if an earlier test fails
  - Each `RUN_TEST()` should print a test result. This could be parsed by the runner to generate a test report of some kind.

Reply:
Hi @Dazza0,
To avoid impacting between different test cases, it uses individual images for each test case. Which means it has serveral images for every test case. Each test case has its own test runners. That's why it calls `UNITY_BEGIN()`/`RUN_TEST()`/`UNITY_END()` individually.